### PR TITLE
Removes decca and refactors Permissions into React

### PR DESCRIPTION
**Pull request by josemigallas**
**What this PR does / why we need it**:

This PR aims to remove old and obsolete libraries such as `decca` and refactor the whole Permissions module into React components, also reviewing specs and general implementation.

At a first step it removes `decca`, `classnames` and refactors `Permissions/permissions.js` into `Permissions/permissions.jsx`, ditching the current custom implementation of a state in favor of built-in React's.

**Which issue(s) this PR fixes** 
This is a rogue PR, no issues involved.

**Verification steps** 

1. Go to any user and edit (/account/users/[user-id]/edit)
2. Try updating with different configurations each time: all services, no services, some services, etc.

### CODE REVIEW

File: app/javascript/packs/permissions.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import { render as renderPermissionsWidget } from 'Users/permissions'
+import { PermissionsFormWrapper } from 'Users/components/PermissionsForm'
```
Comment (damianpm) [inline app/javascript/packs/permissions.js]: PermissionsForm file is missing

File: spec/javascripts/Users/components/PermissionsForm.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    wrapper.find('.user_member_permission_ids').at(1).simulate('change')
+    expect(wrapper.state('checkedFeatures')).toHaveLength(1)
+    wrapper.find('.user_member_permission_ids').at(2).simulate('change')
+    expect(wrapper.state('checkedFeatures')).toHaveLength(2)
+    wrapper.find('.user_member_permission_ids').at(3).simulate('change')
+    expect(wrapper.state('checkedFeatures')).toHaveLength(3)
+  })
+
+  it.skip('should render a ServiceFeatureAccess if any feature granting service access is checked', () => {
```
Comment (damianpm) [inline spec/javascripts/Users/components/PermissionsForm.spec.jsx]: Why skip?

File: app/javascript/src/Users/components/FeatureAccess.jsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+type Props = {
+  value: string,
+  checked?: boolean,
+  onChange: string => void,
+  children?: React.Node
+}
+
+const FeatureAccess = ({ value, checked = false, onChange, children }: Props) => {
+  let liClass = `FeatureAccessList-item FeatureAccessList-item--${value} ${checked ? 'is-checked' : 'is-unchecked'}`
+  let inputClass = `user_member_permission_ids ${isServicePermissionsGranted(value) ? 'user_member_permission_ids--service' : ''}`
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply app/javascript/src/Users/components/FeatureAccess.jsx]: It's how styles are written unfortunately

File: app/javascript/src/Users/components/FeatureAccessInput.jsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+const FeatureAccessInput = ({ showServices = false, children }: Props) => {
+  const olClass = `FeatureAccessList ${showServices ? '' : 'FeatureAccessList--noServicePermissionsGranted'}`
+
+  return (
+    <fieldset>
+      <legend className='label'><label>This user can access</label></legend>
+      <ol className={olClass}>
+        {children}
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply app/javascript/src/Users/components/FeatureAccessInput.jsx]: I guess it'd make more sense to make `children` not optional instead.

File: package.json [hunk 5]
```diff
@@ -97,6 +95,7 @@
     "redux-thunk": "^2.3.0",
     "select2": "^4.0.6-rc.1",
     "url-polyfill": "^1.1.3",
+    "virtual-dom": "^2.1.1",
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply package.json]: This dependency was included indirectly installed with `decca` and it's used by `Stats`.

File: spec/javascripts/Users/components/AdminSection.spec.jsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+afterEach(() => {
+  wrapper.unmount()
+})
+
+it('should render itself', () => {
+  expect(wrapper.find(AdminSection).exists()).toBe(true)
+})
+
+it('should render its children', () => {
```
[the remark this answers is not in the document]
Comment (josemigallas) [reply spec/javascripts/Users/components/AdminSection.spec.jsx]: With no children, it will render nothing. This one is just to assert composition.

File: spec/javascripts/Users/components/PermissionsForm.spec.jsx (hunk 2 above)
Comment (josemigallas) [reply spec/javascripts/Users/components/PermissionsForm.spec.jsx]: Well the spec is not yet implemented, it's a to-do for a future PR.

File: app/javascript/src/Users/components/AdminSection.jsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import * as React from 'react'
+
+type Props = {
+  name: string,
+  available?: boolean,
+  children?: React.Node
+}
+
+const AdminSection = ({ name, available = false, children }: Props) => {
+  let adminSectionClass = `ServiceAccessList-sectionItem ServiceAccessList-sectionItem--${name} ${available ? '' : 'is-unavailable'}`
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/AdminSection.jsx]: you are not planning to reassign `adminSectionClass`, right?

File: app/javascript/src/Users/components/FeatureAccess.jsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+type Props = {
+  value: Feature,
+  checked?: boolean,
+  onChange: Feature => void,
+  children?: React.Node
+}
+
+const FeatureAccess = ({ value, checked = false, onChange, children }: Props) => {
+  let liClass = `FeatureAccessList-item FeatureAccessList-item--${value} ${checked ? 'is-checked' : 'is-unchecked'}`
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeatureAccess.jsx]: You could use `const` here as well.

File: app/javascript/src/Users/components/PermissionsForm.jsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+type Props = {
+  initialState: any,
+  features: Feature[],
+  services: Service[]
+}
+
+type State = {
+  role: Role,
+  checkedFeatures: Array<Feature | 'services'>,
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/PermissionsForm.jsx]: This looks weird, having an `Array` of `Feature` or `string services`?

File: app/javascript/src/Users/components/PermissionsForm.jsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  checkedServicesIds: number[]
+}
+
+class PermissionsForm extends React.Component<Props, State> {
+  state = {
+    role: this.props.initialState.role || 'admin',
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/PermissionsForm.jsx]: we are passing the initialState as props? Is there a way to craft the initial state from the props? we also will be reducing the coupling.

File: app/javascript/src/Users/components/PermissionsForm.jsx [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    checkedFeatures: this.props.initialState.admin_sections || [],
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/PermissionsForm.jsx]: Maybe when defining the initial state in a separated fn, we could assign the default values. Still is not bad, I know it's widely used this way when defining the state on class components. I'll let you to decide.

File: app/javascript/src/Users/components/PermissionsForm.jsx [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    checkedServicesIds: this.props.initialState.member_permission_service_ids || []
+  }
+
+  handleRoleChange = (role: Role) => this.setState({ role })
+
+  handleFeatureChecked = (feature: Feature | 'services') => {
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/PermissionsForm.jsx]: Is the string `services` a type `Feature` ?

File: app/javascript/src/Users/utils/index.jsx [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+export function getFeatureName (feature: Feature): string {
+  if (feature in FEATURE_NAMES) {
+    return FEATURE_NAMES[feature]
+  }
+
+  throw new Error(`${feature} is not a known feature`)
+}
+
+export function isServicePermissionsGranted (features: Feature | 'services' | Array<Feature | 'services'>): boolean {
```
Comment (didierofrivia) [inline app/javascript/src/Users/utils/index.jsx]: yeah, It's a bit weird how the string `services` mingles with Features, maybe define a type for that string?

File: app/javascript/src/Users/utils/index.jsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  if (typeof features === 'string') {
```
Comment (didierofrivia) [inline app/javascript/src/Users/utils/index.jsx]: and here we could match for a specific type

File: app/javascript/src/Users/utils/index.jsx [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  } else {
+    return !!features.find(isServicePermissionsGranted)
+  }
+}
+
+export function toggleElementInCollection<T> (el: T, collection: T[]): T[] {
+  const i = collection.indexOf(el)
+
+  if (i > -1) {
+    collection.splice(i, 1)
```
Comment (didierofrivia) [inline app/javascript/src/Users/utils/index.jsx]: This will be used to set a new state, right? Maybe we could avoid modifying the collection and returning a new Array. I feel uneasy when I see `splice` and `push` methods hehe

File: app/views/provider/admin/account/users/_form.html.erb [hunk 16]
```diff
@@ -57,11 +57,10 @@
         <script>
           document.addEventListener('DOMContentLoaded', function () {
```
Comment (didierofrivia) [inline app/views/provider/admin/account/users/_form.html.erb]: we could use ruby `content_for` here and pass them as `data` attr, then initialize the `permissionsWidget` inside the pack.

File: app/javascript/src/Users/components/RoleInput.jsx [hunk 17]
```diff
@@ -0,0 +1,18 @@
+import * as React from 'react'
+
+type Props = {
+  children?: React.Node
+}
+
+const RoleInput = ({ children }: Props) => (
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/RoleInput.jsx]: Tiny components like `AdminSection`, `FeatureAccessInput`, etc... IMHO, don't deserve to be in a separated file since they are used only by one Component... It's a bit annoying going to another file to see the definition of something that could've been defined in the same file, given the impact of these submodules. Anyway, I know your preferences

Comment (josemigallas) [reply app/javascript/src/Users/components/RoleInput.jsx]: The main reasons why I do this are tidiness: components are easier to locate by using the project explorer or even using a _go to file..._ feature using the components name; and readability: having less lines of code makes the component faster to read and easier to understand.

**But you're right, it's a bit overkill with such small components**. I will refactor it and find a better way to group them.

File: app/javascript/src/Users/components/PermissionsForm.jsx [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+import { Inputs } from 'Users/components/Inputs'
+import { RoleInput } from 'Users/components/RoleInput'
+import { UserRole } from 'Users/components/UserRole'
+import { FeatureAccessInput } from 'Users/components/FeatureAccessInput'
+import { FeatureAccess } from 'Users/components/FeatureAccess'
+import { ServiceFeatureAccess } from 'Users/components/ServiceFeatureAccess'
+import { ServiceAccessList } from 'Users/components/ServiceAccessList'
+import { ServiceAccess } from 'Users/components/ServiceAccess'
+
```
Comment (damianpm) [inline app/javascript/src/Users/components/PermissionsForm.jsx]: Maybe you can import everything in a single import statement from an index file in `Users/components/index.js`

File: app/javascript/src/Users/components/RoleInput.jsx (hunk 17 above)
Comment (damianpm) [reply app/javascript/src/Users/components/RoleInput.jsx]: IMO, if the piece of code is used ( or is expected to be used) in other files or is too large, it deserves to be a new component. If used in only one file, is not needed.

File: app/javascript/src/Users/types/index.jsx [hunk 19]
```diff
@@ -0,0 +1,4 @@
+// @flow
+
+export * from 'Users/types/Role'
+export * from 'Users/types/Feature'
```
Comment (damianpm) [inline app/javascript/src/Users/types/index.jsx]: Do we need 2 files `Feature.jsx`and `Role.jsx`? Maybe is too much, specially for a single-line like Role.jsx

File: spec/javascripts/Users/components/FeatureAccess.spec.jsx [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+import Enzyme, { mount } from 'enzyme'
+import Adapter from 'enzyme-adapter-react-16'
+
+import { FeatureAccess } from 'Users/components/FeatureAccess'
+
+Enzyme.configure({ adapter: new Adapter() })
+
+let wrapper
+
+function getWrapper (testProps) {
```
Comment (damianpm) [inline spec/javascripts/Users/components/FeatureAccess.spec.jsx]: You are not passing `testProps` in the scope of this file

File: spec/javascripts/Users/components/PermissionsForm.spec.jsx [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+
+import React from 'react'
+import Enzyme, { mount } from 'enzyme'
+import Adapter from 'enzyme-adapter-react-16'
+
+import { PermissionsForm } from 'Users/components/PermissionsForm'
+
+Enzyme.configure({ adapter: new Adapter() })
+
+function getWrapper (testProps) {
```
Comment (damianpm) [inline spec/javascripts/Users/components/PermissionsForm.spec.jsx]: Same here, `testProps` not being passed

File: spec/javascripts/Users/components/ServiceAccessList.spec.jsx [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+import Enzyme, { mount } from 'enzyme'
+import Adapter from 'enzyme-adapter-react-16'
+
+import { ServiceAccessList } from 'Users/components/ServiceAccessList'
+
+Enzyme.configure({ adapter: new Adapter() })
+
+let wrapper
+
+function getWrapper (testProps) {
```
Comment (damianpm) [inline spec/javascripts/Users/components/ServiceAccessList.spec.jsx]: And here

File: spec/javascripts/Users/components/ServiceFeatureAccess.spec.jsx [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+import Enzyme, { mount } from 'enzyme'
+import Adapter from 'enzyme-adapter-react-16'
+
+import { ServiceFeatureAccess } from 'Users/components/ServiceFeatureAccess'
+
+Enzyme.configure({ adapter: new Adapter() })
+
+let wrapper
+
+function getWrapper (testProps) {
```
Comment (damianpm) [inline spec/javascripts/Users/components/ServiceFeatureAccess.spec.jsx]: And here

File: app/javascript/src/Users/components/PermissionsForm.jsx (hunk 9 above)
Comment (josemigallas) [reply app/javascript/src/Users/components/PermissionsForm.jsx]: I agree it looks weird but this is how our API is defined. The array contains all *checked features* plus it may contain the term *services* that is not a feature. When it is _not_ present, it means the whole service set is enabled. When it _is_ present, it means services have been enabled manually and some may not been included. See following pictures:

Screen Shot 2019-04-02 at 16 55 02
>`admin_sections` contains only features (billing, settings, analytics, etc) and all services are enabled. `member_permission_service_ids` is `null`.

Screen Shot 2019-04-02 at 16 55 36
>`admin_sections` here contains the term `services` that indicates a set of services is specified in `member_permissions_service_ids`.

File: app/javascript/src/Users/components/PermissionsForm.jsx (hunk 10 above)
Comment (josemigallas) [reply app/javascript/src/Users/components/PermissionsForm.jsx]: The state is being crafted from the props, it just happens the props is called `initialState` as [it was `state` before](https://github.com/3scale/porta/blob/master/app/views/provider/admin/account/users/_form.html.erb#L61). Do you think we should rename the props?

File: app/javascript/src/Users/components/PermissionsForm.jsx (hunk 12 above)
Comment (josemigallas) [reply app/javascript/src/Users/components/PermissionsForm.jsx]: Clarification for this above https://github.com/3scale/porta/pull/604#discussion_r271355724

File: app/javascript/src/Users/utils/index.jsx (hunk 14 above)
Comment (josemigallas) [reply app/javascript/src/Users/utils/index.jsx]: But it matches for a specific type: `string`, otherwise `array`.
Grouping Feature and 'service' in a type would end up being more convoluted:
```
isServicePermissionsGranted (features: FeatureOrServices | Array<FeatureOrServices>) {
  if (typeof features === FeatureOrServices) {
     ...
```
It reads almost the same and keep in mind `Feature` is just a collection of strings, not a type per se.

File: spec/javascripts/Users/components/FeatureAccess.spec.jsx (hunk 20 above)
Comment (josemigallas) [reply spec/javascripts/Users/components/FeatureAccess.spec.jsx]: Well it's a test template. `getWrapper` is prepared for whenever `testProps` are needed.

File: app/javascript/src/Users/components/PermissionsForm.jsx (hunk 9 above)
Comment (didierofrivia) [reply app/javascript/src/Users/components/PermissionsForm.jsx]: What about defining a type `AdminSections` where it will be a list of `string` and then a `UserAdminSections a` where `a` would be a List `AdminSections`. Also, you could keep a type `Services` in your state among the role and serviceCheckedIds. Then `Features` defined somewhere (you don't care about them in the state of your app). Then you could decide if a Feature is checked based on the admiSections you've got from your initialState. That way would be type bulletproof

File: spec/javascripts/Users/components/FeatureAccess.spec.jsx (hunk 20 above)
Comment (damianpm) [reply spec/javascripts/Users/components/FeatureAccess.spec.jsx]: If this were a method imported from somewhere else, it would have sense. But if you are declaring the same function in every file it doesn't look like a template anymore.

**Comment by josemigallas** [comment]
@didierofrivia @damianpm 
Trying with some new approaches, let me know what you think. I'd rather have the smaller components by far, but I understand if everyone else don't like it. Some other changes involve using `useState` and adding a new type `AdminSections`. 👷**Keep in mind it's a WIP**!

File: app/javascript/src/Policies/util.jsx [hunk 24]
```diff
@@ -48,28 +48,12 @@ function parsePolicy (key: string, policy: RawPolicy): RegistryPolicy {
   return { ...policy, name: key, humanName: policy.name, data: {} }
 }
 
-const toJsonString = (val: Object): string => JSON.stringify(val, null, 2)
```
Comment (didierofrivia) [inline app/javascript/src/Policies/util.jsx]: Could you extract this to a different PR?

Comment (didierofrivia) [reply app/javascript/src/Policies/util.jsx]: The extracting to `utiliities/json-utils` and the import in the other files

File: app/javascript/src/Users/components/FeatureAccessInput.jsx [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+
+type Props = {
+  features: Feature[],
+  checkedAdminSections: AdminSections[],
+  showServices?: boolean,
+  onChange: AdminSections => void
+}
+
+const FeatureAccessInput = ({ features, checkedAdminSections, showServices = false, onChange }: Props) => {
+  const olClass = `FeatureAccessList ${showServices ? '' : 'FeatureAccessList--noServicePermissionsGranted'}`
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeatureAccessInput.jsx]: `FeatureAccessListClass`?

File: app/javascript/src/Users/components/FeatureAccessInput.jsx [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+import type { Feature, AdminSections } from 'Users/types'
+
+type Props = {
+  features: Feature[],
+  checkedAdminSections: AdminSections[],
+  showServices?: boolean,
+  onChange: AdminSections => void
+}
+
+const FeatureAccessInput = ({ features, checkedAdminSections, showServices = false, onChange }: Props) => {
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeatureAccessInput.jsx]: This is not just an input, maybe renaming to something more meaningful?

File: app/javascript/src/Users/components/FeatureAccessInput.jsx (hunk 25 above)
Comment (didierofrivia) [reply app/javascript/src/Users/components/FeatureAccessInput.jsx]: I see you use it in other places the same way, denoting just the html structure related class. It's only good when there's only one element of each, either way, better to use specific names.

File: app/javascript/src/Users/components/FeatureAccessInput.jsx [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+  )
+}
+
+const FeatureAccess = ({ value, checked = false, onChange, children }: {
+  value: Feature,
+  checked?: boolean,
+  onChange: Feature => void,
+  children?: React.Node
+}) => {
+  const liClass = `FeatureAccessList-item FeatureAccessList-item--${value} ${checked ? 'is-checked' : 'is-unchecked'}`
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeatureAccessInput.jsx]: Maybe then, `liClassSet` or `liClassNames` or `liClasses`. Same for the rest

File: app/javascript/src/Users/components/FeatureAccessInput.jsx [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+      </ol>
+    </fieldset>
+  )
+}
+
+const FeatureAccess = ({ value, checked = false, onChange, children }: {
+  value: Feature,
+  checked?: boolean,
+  onChange: Feature => void,
+  children?: React.Node
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeatureAccessInput.jsx]: I'd rather request here an argument `featureName` of type `string` since this is what is really needed in this component, right?

File: app/javascript/src/Users/components/FeatureAccessInput.jsx [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+          checked={checked}
+          onChange={() => onChange(value)}
+        />
+        {children}
+      </label>
+    </li>
+  )
+}
+
+const ServiceFeatureAccess = ({ checked = false, onChange, children }: {
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeatureAccessInput.jsx]: `ServiceAccessFeature` maybe?

File: app/javascript/src/Users/utils/index.jsx [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+
+export function isServicePermissionsGranted (features: Feature | 'services' | Array<Feature | 'services'>): boolean {
+  if (typeof features === 'string') {
+    return FEATURES_GRANTING_SERVICE_ACCESS.includes(features)
+  } else {
+    return !!features.find(isServicePermissionsGranted)
+  }
+}
+
+export function toggleElementInCollection<T> (el: T, collection: T[]): T[] {
```
Comment (didierofrivia) [inline app/javascript/src/Users/utils/index.jsx]: What about ...
```
const i = collection.indexOf(el)
return (i > -1) ? [...collection, el] : collection.filter((_, j) => j != i)
```
?

**Comment by didierofrivia** [comment]
I kept in mind is a WIP, I like the new direction. Left a few comments, will resume the review when you advance more on this. Thanks!

File: app/javascript/src/Users/components/FeatureAccessInput.jsx (hunk 28 above)
Comment (josemigallas) [reply app/javascript/src/Users/components/FeatureAccessInput.jsx]: Right, having children here does not make sense at all.

File: app/javascript/src/Policies/util.jsx (hunk 24 above)
Comment (josemigallas) [reply app/javascript/src/Policies/util.jsx]: Done #746

File: app/javascript/src/utilities/json-utils.jsx [hunk 31]
```diff
@@ -0,0 +1,22 @@
+// @flow
+
+// eslint-disable-next-line flowtype/no-weak-types
+const toJsonString = (val: Object): string => JSON.stringify(val, null, 2)
+
+const fromJsonString = (json: string) => JSON.parse(json)
+
+const safeFromJsonString = (json: string) => {
+  console.log(json)
```
Comment (damianpm) [inline app/javascript/src/utilities/json-utils.jsx]: Are you deleting this?

File: app/javascript/src/utilities/json-utils.jsx [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+  try {
+    return fromJsonString(json)
+  } catch (err) {
+    console.warn('That doesn\'t look like a valid json!', err)
```
Comment (damianpm) [inline app/javascript/src/utilities/json-utils.jsx]: Why a warn and not throwing an error?

File: app/javascript/src/Users/utils/index.jsx (hunk 30 above)
Comment (josemigallas) [reply app/javascript/src/Users/utils/index.jsx]: It's less readable IMO, also it takes twice as much time in Firefox, about the same time in Chrome. I like it but it uses complex syntax for such a small method.

File: app/javascript/src/Users/components/FeaturesFieldset.jsx [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+type Props = {
+  features: Feature[],
+  selectedSections?: AdminSections[],
+  areServicesVisible?: boolean,
+  onAdminSectionSelected: AdminSections => void
+}
+
+/**
+ * Contains a set of Features that can be checked by the user.
+ * @param {Feature[]} features - An array containing all available Features in the fieldset.
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeaturesFieldset.jsx]: Nice that you are adding documentation, don't stop there and space them so they are aligned and easy to read ;)

File: app/javascript/src/Users/components/ServicesFieldset.jsx [hunk 34]
```diff
@@ -0,0 +1,74 @@
...
+import type { Service } from 'Types'
+import type { AdminSections } from 'Users/types'
+
+const ServicesFieldset = ({ services = [], selectedSections = [], selectedServicesIds = [], onServiceSelected }: {
+  services?: Service[],
+  selectedSections?: AdminSections[],
+  selectedServicesIds?: number[],
+  onServiceSelected: number => void
+}) => {
+  const servicesList = `ServiceAccessList`
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/ServicesFieldset.jsx]: A problem that I see with this const name, is that I might be expecting an Array of services because of the "List" part. And this is just a css class name. I'd rather name it `serviceListClassName` or we could agree on something like `serviceListCss` or even all related css classes as `cssNameOfTheSet`. Also, use single quotes not template string. And to simplify it more, in this case, we might not even need a const to hold that sole value.

File: app/javascript/src/Users/components/FeaturesFieldset.jsx [hunk 35]
```diff
@@ -0,0 +1,107 @@
...
+ * @param {Feature} value - The feature this checkbox represents.
+ * @param {boolean} checked - Whether or not this checkbox is selected.
+ * @param {Function} onChange - A callback function triggered when its value changes.
+ */
+const FeatureCheckbox = ({ value, checked, onChange }: {
+  value: Feature,
+  checked?: boolean,
+  onChange: Feature => void
+}) => {
+  const featuresListItem = `FeatureAccessList-item FeatureAccessList-item--${value} ${checked ? 'is-checked' : 'is-unchecked'}`
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/FeaturesFieldset.jsx]: Here I'd make it obvious that it's a list of classNames. This is related to https://github.com/3scale/porta/pull/604/files#r272902142

File: app/javascript/src/Users/components/PermissionsForm.jsx [hunk 36]
```diff
@@ -0,0 +1,77 @@
...
+    member_permission_service_ids?: number[]
+  },
+  features: Feature[],
+  services: Service[]
+}
+
+const PermissionsForm = ({ initialState = {}, features, services }: Props) => {
+  const [role, setRole] = useState(initialState.role || 'admin')
+  const [selectedSections, setSelectedSections] = useState(initialState.admin_sections || [])
+  const [selectedServicesIds, setSelectedServicesIds] = useState(initialState.member_permission_service_ids || [])
```
Comment (didierofrivia) [inline app/javascript/src/Users/components/PermissionsForm.jsx]: And yeah, I prefer the new hooks approach :+1:

File: app/javascript/src/Users/types/index.jsx [hunk 37]
```diff
@@ -0,0 +1,13 @@
...
+
+export type Feature = 'portal' |
+'finance' |
+'settings' |
+'partners' |
+'monitoring' |
+'plans' |
+'policy_registry'
+
+export type AdminSections = Feature | 'services'
```
Comment (didierofrivia) [inline app/javascript/src/Users/types/index.jsx]: maybe better `AdminSection`, singular?

File: app/javascript/src/Users/utils/index.jsx [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+
+export function getFeatureName (feature: Feature): string {
+  if (feature in FEATURE_NAMES) {
+    return FEATURE_NAMES[feature]
+  }
+
+  throw new Error(`${feature} is not a known feature`)
+}
+
+export function canFeatureSetServicePermissions (features: AdminSections | Array<AdminSections>): boolean {
```
Comment (didierofrivia) [inline app/javascript/src/Users/utils/index.jsx]: Now here having `AdminSections` and ` Array<AdminSections>` suggest a multidimensional Array. Could be fixed with https://github.com/3scale/porta/pull/604/files#r272903513

File: app/javascript/src/Users/components/ServicesFieldset.jsx (hunk 34 above)
Comment (josemigallas) [reply app/javascript/src/Users/components/ServicesFieldset.jsx]: I wasn't sure either, I wanted t avoid repetition but I agree it is misleading. I'll try with the ClassName at the end.

File: app/javascript/src/Users/utils/index.jsx (hunk 30 above)
Comment (didierofrivia) [reply app/javascript/src/Users/utils/index.jsx]: IMHO, it's more elegant and the benchmark you used it's not reliable, since it would depend on the addons and settings of your browsers (and possibly the resources in use in your machine). Also, it's just in the order of nanoseconds, it's not used in a non tail recursive operation or any in that case and it's just 2 lines of code against 10. But since this is not at all a dealbreaker to me, I'll leave it sink in your mind and/or maybe @damianpm or anyone else can give their opinion?

Comment (damianpm) [reply app/javascript/src/Users/utils/index.jsx]: I'd prefer the shorter syntax. It's true it's more cryptic, but you reduce to only 2 lines

Comment (josemigallas) [reply app/javascript/src/Users/utils/index.jsx]: _Shorter + cryptic_ vs. _longer + descriptive_, I think we all know who's the winner. But since it's already 2 against me I won't hold this any longer.

Comment (didierofrivia) [reply app/javascript/src/Users/utils/index.jsx]: suck it up, and carry on...

File: spec/javascripts/Users/components/FeaturesFieldset.spec.jsx [hunk 39]
```diff
@@ -0,0 +1,116 @@
...
+
+afterEach(() => {
+  wrapper.unmount()
+})
+
+it('should render itself', () => {
+  expect(wrapper.find(FeaturesFieldset).exists()).toBe(true)
+})
+
+it('should render a hidden input', () => {
```
Comment (didierofrivia) [inline spec/javascripts/Users/components/FeaturesFieldset.spec.jsx]: This hidden input should have an specific prop, right?

File: spec/javascripts/Users/components/PermissionsForm.spec.jsx [hunk 40]
```diff
@@ -0,0 +1,235 @@
+// @flow
```
Comment (didierofrivia) [inline spec/javascripts/Users/components/PermissionsForm.spec.jsx]: It's no needed for specs. Up to you ;)

Comment (didierofrivia) [reply spec/javascripts/Users/components/PermissionsForm.spec.jsx]: we might want to benchmark the time of execution for the tests with and w/o it.

Comment (didierofrivia) [reply spec/javascripts/Users/components/PermissionsForm.spec.jsx]: I'd say we should remove it, and also remove it from webpacker for our test folder. The latter can be done in a different PR, since it doesn't add any value to our tests, only time. Our flow already has a specific task.

File: package-lock.json [hunk 41]
```diff
@@ -5,7 +5,7 @@
     "@babel/code-frame": {
       "version": "7.0.0",
       "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.0.0.tgz",
-      "integrity": "sha512-OfC2uemaknXr87bdLUkWog7nYuliM9Ij5HUcajsVcMCpQrcLmtxRbVFTIqmcSkSeYRBFBRxs2FiUqFJDLdiebA==",
+      "integrity": "sha1-BuKrGb21NThVWaq7W6WXKUgoAPg=",
```
Comment (mikz) [inline package-lock.json]: looks like you used some older npm?

Comment (josemigallas) [reply package-lock.json]: `6.9.0` in fact 😕
